### PR TITLE
Music album details

### PR DESCRIPTION
**Pull request by nielsvanvelzen**
Depends on #443

### CODE REVIEW

**Comment by nielsvanvelzen** [comment]
Some additional context. I took the layout files of the `ItemListActivity` but converted it to use rows / presenters . The layout itself is mostly the same as I didn't want to redo that right now. The way it is made should make it easy to reuse the list for episodes in a season view.

**Comment by AndreasGB** [comment]
I found a bug. When you have a genre selected and then scroll down to the first track, it is scrolled off screen.

image
Press down once
image

**Comment by nielsvanvelzen** [comment]
I think this is because your list is too long, I'll try to get a album in my test library with 15+ songs to test it!

**Comment by AndreasGB** [comment]
Quick fix via pop-up: "Your album has x surplus songs, please delete them."

**Comment by nielsvanvelzen** [comment]
Turns out it's not that easy to fix so I'll need to find some time to have a good look at it. Marking this as draft for now.

**Comment by nielsvanvelzen** [comment]
(Build fails because of an issue already fixed on master, see #474. Please ignore that when reviewing/merging)

**Comment by anthonylavado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346&bid=18135265)

File: app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.jellyfin.androidtv.R
+import org.jellyfin.androidtv.TvApp
+import org.jellyfin.androidtv.model.itemtypes.Audio
+import org.jellyfin.androidtv.playback.MediaManager
+import org.jellyfin.androidtv.util.TimeUtils
+import org.jellyfin.androidtv.util.Utils
+import org.jellyfin.androidtv.util.apiclient.PlaybackHelper
+import org.jellyfin.androidtv.util.apiclient.getItem
+
+class SongPresenter : Presenter() {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt:22]: Codacy Issue found: [UndocumentedPublicClass - [SongPresenter] at /src/app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt:22:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt [hunk 2]
```diff
@@ -134,11 +145,13 @@ private suspend fun ApiClient.getEpisodesOfSeason(seasonId: String): List<Episod
 	})
 }
 
-suspend fun ApiClient.getAlbumsForArtist(artist: Artist): List<Album>? = suspendCoroutine { continuation ->
+suspend fun ApiClient.getAlbumsForArtist(artist: Artist): List<Album>? = getAlbumsForArtists(arrayOf(artist.id))
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:148]: Codacy Issue found: [UndocumentedPublicFunction - [getAlbumsForArtist] at /src/app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:148:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/details/fragments/AlbumDetailsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import org.jellyfin.androidtv.details.actions.PlayFromBeginningAction
+import org.jellyfin.androidtv.details.actions.ShuffleAction
+import org.jellyfin.androidtv.details.actions.ToggleFavoriteAction
+import org.jellyfin.androidtv.details.presenters.ItemPresenter
+import org.jellyfin.androidtv.details.presenters.SongPresenter
+import org.jellyfin.androidtv.details.rows.DetailsOverviewRow
+import org.jellyfin.androidtv.model.itemtypes.Album
+import org.jellyfin.androidtv.util.addIfNotEmpty
+import org.jellyfin.androidtv.util.apiclient.getAlbumsForArtists
+import org.jellyfin.androidtv.util.apiclient.getSimilarItems
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/fragments/AlbumDetailsFragment.kt:22]: Codacy Issue found: [UndocumentedPublicClass - [AlbumDetailsFragment] at /src/app/src/main/java/org/jellyfin/androidtv/details/fragments/AlbumDetailsFragment.kt:26:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt [hunk 4]
```diff
@@ -86,10 +87,21 @@ class LocalTrailer(original: BaseItemDto) : PlayableItem(original)
...
+	override val criticsRating: Float? = original.criticRating
+}
 
 class Artist(original: BaseItemDto) : BaseItem(original), Ratable {
 	override val officialRating: String? = original.officialRating
 	override val communityRating: Float? = original.communityRating
 	override val criticsRating: Float? = original.criticRating
 }
+
+class Audio(original: BaseItemDto) : PlayableItem(original) {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt:104]: Codacy Issue found: [UndocumentedPublicClass - [Audio] at /src/app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt:104:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		val rowItem = item as VerticalListRow
+
+		vh.itemBridgeAdapter.setAdapter(rowItem.adapter)
+		vh.items.apply {
+			layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
+			adapter = vh.itemBridgeAdapter
+		}
+	}
+
+	class ViewHolder(view: View) : RowPresenter.ViewHolder(view) {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt:40]: Codacy Issue found: [UndocumentedPublicClass - [ViewHolder] at /src/app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt:40:2](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt [hunk 6]
```diff
@@ -153,3 +166,23 @@ suspend fun ApiClient.getAlbumsForArtist(artist: Artist): List<Album>? = suspend
 		}
 	})
 }
+suspend fun ApiClient.getSongsForAlbum(albumId: String): List<Audio>? = suspendCoroutine { continuation ->
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:169]: Codacy Issue found: [UndocumentedPublicFunction - [getSongsForAlbum] at /src/app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:169:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+			menu.add(R.string.lbl_instant_mix).setOnMenuItemClickListener {
+				PlaybackHelper.playInstantMix(item.id)
+
+				true
+			}
+		}.show()
+	}
+
+	class ViewHolder(view: View) : Presenter.ViewHolder(view) {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt:78]: Codacy Issue found: [UndocumentedPublicClass - [ViewHolder] at /src/app/src/main/java/org/jellyfin/androidtv/details/presenters/SongPresenter.kt:78:2](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/details/rows/VerticalListRow.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package org.jellyfin.androidtv.details.rows
+
+import androidx.leanback.widget.HeaderItem
+import androidx.leanback.widget.ListRow
+import androidx.leanback.widget.ObjectAdapter
+
+class VerticalListRow(
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/rows/VerticalListRow.kt:7]: Codacy Issue found: [UndocumentedPublicClass - [VerticalListRow] at /src/app/src/main/java/org/jellyfin/androidtv/details/rows/VerticalListRow.kt:7:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt [hunk 9]
```diff
@@ -134,11 +145,13 @@ private suspend fun ApiClient.getEpisodesOfSeason(seasonId: String): List<Episod
 	})
 }
 
-suspend fun ApiClient.getAlbumsForArtist(artist: Artist): List<Album>? = suspendCoroutine { continuation ->
+suspend fun ApiClient.getAlbumsForArtist(artist: Artist): List<Album>? = getAlbumsForArtists(arrayOf(artist.id))
+
+suspend fun ApiClient.getAlbumsForArtists(artists: Array<String>): List<Album>? = suspendCoroutine { continuation ->
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:150]: Codacy Issue found: [UndocumentedPublicFunction - [getAlbumsForArtists] at /src/app/src/main/java/org/jellyfin/androidtv/util/apiclient/ApiClientExtensions.kt:150:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import android.view.View
+import android.view.ViewGroup
+import androidx.leanback.widget.ItemBridgeAdapter
+import androidx.leanback.widget.RowPresenter
+import androidx.recyclerview.widget.LinearLayoutManager
+import kotlinx.android.synthetic.main.row_vertical_list.view.*
+import org.jellyfin.androidtv.R
+import org.jellyfin.androidtv.details.rows.VerticalListRow
+
+class VerticalListRowPresenter : RowPresenter() {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt:13]: Codacy Issue found: [UndocumentedPublicClass - [VerticalListRowPresenter] at /src/app/src/main/java/org/jellyfin/androidtv/details/presenters/VerticalListRowPresenter.kt:13:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

File: app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt [hunk 11]
```diff
@@ -86,10 +87,21 @@ class LocalTrailer(original: BaseItemDto) : PlayableItem(original)
 
 class Video(original: BaseItemDto) : PlayableItem(original)
 
-class Album(original: BaseItemDto) : BaseItem(original)
+class Album(original: BaseItemDto) : BaseItem(original), Ratable {
```
Comment (anthonylavado) [inline app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt:90]: Codacy Issue found: [UndocumentedPublicClass - [Album] at /src/app/src/main/java/org/jellyfin/androidtv/model/itemtypes/LiftedItems.kt:90:1](https://app.codacy.com/gh/jellyfin/jellyfin-androidtv/pullRequest?prid=5532346)

**Comment by thornbill** [comment]
It doesn't seem like the buttons in the action row (play, shuffle, etc.) are working. Is that expected?

**Comment by nielsvanvelzen** [comment]
They should work... Maybe I did a bad rebase, I'll look at it

**Pull request merged**